lib/utils/metadata.js

```diff
@@ -0,0 +1,366 @@
+ * governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+const gm = require("./gm-promisify");
+const unitConvert = require("css-unit-converter");
+const util = require('util');
+const child_process = require('child_process');
+const exec = util.promisify(child_process.exec);
+const { GenericError, AssetComputeLogUtils } = require('@adobe/asset-compute-commons');
+const mime = require('mime-types');
+const path = require('path');
+
+const MEDIAINFO_TIMEOUT_MS = 30 * 1000; // max timeout of 30 seconds
+
+// FFMPEG accepted formats list from here: https://git.corp.adobe.com/nui/core/blob/master/lib/workers.js
+const FFMPEG_EXTS = new Set(["avi", "divx", "f4v", "flv", "m2t", "m2ts", "m2v", "m4v", "mkv", "mov", "mp4",
+    "mpeg", "mpg", "mts", "ogv", "qt", "r3d", "webm", "wmv", "3gp", "mxf"
+]);
+
+// Text transcription service accepted formats list from here: https://git.corp.adobe.com/nui/core/blob/master/lib/workers.js
+const TEXT_TRANSCRIPTION_EXTS = new Set(["mp4", "flv", "mxf", "mxf ", "wmv", "asf", "avi", "m4a", "m4v", "wav", "mov", "3gp", "mpg"]);
+
+Object.filter = (obj, predicate) =>
+    Object.keys(obj)
+        .filter(key => predicate(obj[key]))
+        // eslint-disable-next-line no-sequences
+        .reduce((res, key) => (res[key] = obj[key], res), {});
+class Metadata {
+    /**
+     * @param {Object} source 
+     */
+    constructor(source) {
+        if (!source) {
+            throw new GenericError(`No source provided`);
+        } else if (!source.name && !source.path && !source.mimetype) {
+            throw new GenericError(`Provided source must contain one of 'name', 'path', or 'mimetype' fields`);
+        }
+        this._source = source;
+
+        this._extension = this.getExtension(this._source);
+        if (!this._extension) {
+            throw new GenericError(`Cannot determined asset extension`);
+        }
+        this._extracted = false;
+    }
+
+    async extract() {
+        if (this._extracted) {
+            console.warn('metadata has already been extracted');
+            return;
+        }
+        if (this.isVideo() || this.isTextTranscriptable()) {
+            let input = this._source.url;
+            if (!input && this._source.path) {
+                input = this._source.path;
+            }
+            if (!input) {
+                throw new GenericError(`Image source must contain either 'url' or 'path' field for retrieving metadata`);
+            }
+            this._metadata = await this.getAudioVideoMediaMetadata(input);
+        } else { // image asset
+            if (!this._source.path) {
+                throw new GenericError(`Image source must contain 'path' field for retrieving metadata`);
+            }
+            this._metadata = await this.getImageMetadata(this._source.path);
+        }
+        this._extracted = true;
+    }
+
+    get extension() {
+        return this._extension;
+    }
+
+    get metadata() {
+        if (!this._extracted) {
```

Comment:
You could drop `this._extracted` and verify if `this._metadata` is set instead.

Reply:
removed